### DISCUSSION TITLE
Batch Management

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#170] [Ideas]
I notice that the batch management is not very developed in BiblioteQ so I make some

in the books interface new feature "batch add : give the list of isbn-13 and BiblioteQ will integrate the Z39.50, SRU and OpenLibrary servers to complete"

in the books interface new feature "button to ask BiblioteQ to go to the Open Library or Amazon servers to get the missing covers. if isbn10 is valid of course

in the general view - batch deletion : give the list of isbn-13 and BiblioteQ will delete the documents from its database ; if they are not borrowed".

in the table view - borrowing by selection : select the documents BiblioteQ will put them in the borrowed position for the member you will have to select afterwards".

of course, I don't know if it's a hill or a mountain, it's just some ideas of a user that I share with you to see the realization or not :-)

**Comment by textbrowser** [comment]
Group deletions are present. Select the desired items and press the Delete button.

https://user-images.githubusercontent.com/10701156/153635989-efce33a8-2c7f-419b-b91d-21ef04c2e431.mp4

**Comment by textbrowser** [comment]
Input the query (SELECT id, myoid, type FROM book WHERE ID IN (...)) in the Wizard. Select the items and press the Delete button.

**Comment by meteos77** [comment]
it's crazy the difference of vision between a user and a developer
in more pictured : I ask for a hot dish with plate you propose me a choice of sandwich or to make the kitchen :-)

**Reply by textbrowser** [reply]
> it's crazy the difference of vision between a user and a developer in more pictured : I ask for a hot dish with plate you propose me a choice of sandwich or to make the kitchen :-)

It's perception. If one perceives two roles, then yes. One role, the servant. One role, the served. Roles are unpleasant. One can serve themselves if they are educated in doing so. :)

**Reply by meteos77** [reply]
there are still creators and users, these 2 roles are different in many areas
I try to be in my user role as much as possible with suggestions, feedbacks, and my experience as a library manager

you turn this into a servant/servant while we do this by taste. This is probably the image of the kitchen :-)

I can very well spend hours trying to do 1/1000 of your development but the will must be present to do programming
I can very well wait for a software that will provide me with the functionalities that are essential to me

and not as a possible user of BiblioteQ, I have invested myself in tests, translations and suggestions.
In conclusion, at my level I am quite happy with myself and I hope you are also happy with yourself.

**Reply by textbrowser** [reply]
> there are still creators and users, these 2 roles are different in many areas I try to be in my user role as much as possible with suggestions, feedbacks, and my experience as a library manager
> 
> you turn this into a servant/servant while we do this by taste. This is probably the image of the kitchen :-)
> 
> I can very well spend hours trying to do 1/1000 of your development but the will must be present to do programming I can very well wait for a software that will provide me with the functionalities that are essential to me
> 
> and not as a possible user of BiblioteQ, I have invested myself in tests, translations and suggestions. In conclusion, at my level I am quite happy with myself and I hope you are also happy with yourself.

It's a game of quantum chess. How do I please the user while minimizing the amount of repetitive work? While doing so, how do I work while I remain in a state of learning something new? Is this requested process helpful to others? What will I break in the process? How much specialization is expected if there are alternate solutions? I understand your view. Sure. And at times I find the view encouraging. At other times I find myself painting the same room multiple times with the same paint and same brush. There will be variations, but the journey and the result are mundane. Does that make sense?

**Reply by meteos77** [reply]
we are in the discussion zone to exchange

**Reply by meteos77** [reply]
I would prefer to have the sql part more developed in the sense of making it easier to use by users rather than by specialists

**Reply by meteos77** [reply]
_How can I work while remaining in a state of learning something new?_    only you can be the judge of that


weeding in libraries is common (https://en.wikipedia.org/wiki/Weeding_(library))
so adding this feature to a library management software doesn't seem out of line to me

**Comment by meteos77** [comment]
2 cases of mass deletions in the library database

current situation
1st case : the "gardening" done once or twice a year (after having sorted out the documents too old, damaged or that the subject is not borrowed anymore)
the librarian finds himself with a pile of 20 to 100 documents to remove from the database.

- he scans the identifiers (accession_number) of the documents to have a list of the outputs (txt file 1 document per line)
- he puts the software in "output mode" and reads the identifiers file for deletion in the database

- there is another mode which is "scan and delete" but the disadvantage is that there is no history in file

**Comment by meteos77** [comment]
2nd case : the semester change with the documents of the central library
the library receives 500 documents and returns 500 that must be deleted from the database.

- he receives a csv document with the list of the 500 documents to return
- he checks off the documents found on a paper list
- he scans the identifiers (accession_number) of the documents to have a list of the outputs (txt file 1 document per line)
- he puts the software in "output mode" and reads the identifiers file for deletion in the database

for example
[Liste_039_a-sortir-430-notices.txt.tar.gz](https://github.com/textbrowser/biblioteq/files/8053091/Liste_039_a-sortir-430-notices.txt.tar.gz)

**Comment by meteos77** [comment]
BiblioteQ proposes 
- either a "multi clicks" to get a lot of documents even in multiselection it is unrealistic
- or the SQL which is not accessible to the standard user (at least those I know)

so the suggestion was for a standard user to have a kind of "delete from a list" like "import csv".

**Comment by textbrowser** [comment]
Let's say you have to remove 1,000 entries. Or 500. Or 100.

An input method is not optimal. That is 10 * N characters. A selection method is not optimal. Whatever conceivable application in the present or in the future, something or someone has to input 10 * N characters. The request is not ordinary. As such, SQL is best suited. Otherwise, a new panel which is used infrequently is necessary. These tools simplify solutions which are already present but not in forms which are appealing. :)

**Comment by textbrowser** [comment]
![Screenshot_20220212_085832](https://user-images.githubusercontent.com/10701156/153714313-f01fc596-e379-4aab-8329-ad7ccddd82fe.png)

A sample solution if scanning is available. Not included are fancy icons and scanner status. So, the librarian scans items and the items appear in this list. The items can be removed, or returned, or reserved. Combinations of actions too. This is the optimal solution.

**Reply by textbrowser** [reply]
Other actions can be defined. You mentioned group queries of Z39.50. This has been requested years ago.

Anyway, the most important part of this solution is the scanner. This is why I do not entertain additional input lists.

**Comment by meteos77** [comment]
it looks very good to me it's a kind of basket with the different actions at the bottom?

concerning the scanner, there are 2 options
- the universal one you use the webcams and everyone can have a scanner (the hardest to implement)

the handheld : it behaves like a keyboard, it writes the barcode and adds an end of line

**Reply by textbrowser** [reply]
Right, a basket.

**Comment by textbrowser** [comment]
It can include query status if that's the action which was performed. It can include reservation information, images. It's a special panel where the information is and is not in the database. For reservations, it can also include a method for scanning the member or it can coexist with Members Browser. There are many intricate solutions here and this simple interface can become quite powerful. It can also be an example of connecting the scanner to the main window. For example, the basic search can include a method for searching scanned identifiers.  A checkbox there would allow barcodes to be appended to the input.

**Comment by textbrowser** [comment]
So one can scan the items they wish to delete and perform the query in the main window. This is a superset of the new propose panel. Not necessarily delete, but also modify and / or reserve.

The solution relies on a capable scanner and not on new input methods which are tedious in implementation and use. :P

**Reply by meteos77** [reply]
what is a capable scanner?

**Reply by textbrowser** [reply]
One that works and can be bought by ordinary people. Preferably works with Mac, Windows, Linux. Has standard software and hardware interfaces.

**Reply by textbrowser** [reply]
It can scan text and not just barcodes.

**Comment by textbrowser** [comment]
All your desires will be fulfilled with a scanner.

**Reply by textbrowser** [reply]
I want to search for ten books, list them, and reserve them. Scan them and press this button. They'll be listed. Now select them with Ctrl + A and continue.

I want to query 20 books and have those that are found automatically inserted. Scan them, push Query, and Insert.

I want to prune my library. Scan, search, select, delete.

Etc.

**Reply by meteos77** [reply]
it is the dream functionality

**Reply by textbrowser** [reply]
Those are the requests that I like. :P

**Comment by textbrowser** [comment]
For items which do not have identifiers, that's not a separate issue if the scanner can scan text. Scan a title, BQ can automatically suggest the search type. 10 characters? ISBN. 13? ISBN-13. Etc.

**Reply by meteos77** [reply]
it's ambitious text recognition

**Comment by textbrowser** [comment]
Everyone has Android. I'm familiar with writing software in it. There are two routes: off-the-shelf scanner or create a scanner for BQ with USB connectivity.

**Reply by meteos77** [reply]
Everyone has Android ?

**Reply by textbrowser** [reply]
70%?

**Reply by meteos77** [reply]
version ?

**Comment by textbrowser** [comment]
The Android route is somewhat universal but will require more time. The other route is specific to a single model or a model which has standard interfaces.

**Reply by meteos77** [reply]
I don't know if it's a standard but wouldn't D-BUS bring a plus and an interest for the developer?

**Reply by textbrowser** [reply]
How about Windows and Mac?

**Reply by meteos77** [reply]
I thought that if qt had it it was supported by all the platforms that have qt ? [qtdbus](https://doc.qt.io/qt-5/qtdbus-index.html)

general question for my information : 
isn't qt an overlay so that you developers don't have to worry about the layer below?

**Reply by textbrowser** [reply]
D-BUS is for inter-process communications, yes? We need direct access to the device through the native interface. Smart scanners transfer human-readable characters which can be read by an application. They sometimes use end-of-message delimiters.

**Reply by textbrowser** [reply]
Qt is not perfect and there are nuances. The last weirdness I remember was specific to Bluetooth.

**Comment by textbrowser** [comment]
And the 3rd route is to continue entering characters until the end of time. :) OK, there is also a voice possibility. You dictate the text.

**Reply by meteos77** [reply]
the novelties in the libraries which have means are the switching of the bar codes in RFID... it is for the future :-)

**Reply by textbrowser** [reply]
Until they're compromised.

**Comment by textbrowser** [comment]
You've already requested a scanner. Most of your other requests hover around the scanner solution.

**Reply by meteos77** [reply]
the efficiency of scanning is obvious, 
if we can do without mouse (if you put keyboard shortcuts to the button) the human/machine interface will be really great.

**Comment by textbrowser** [comment]
Questions? Thoughts? Concerns?

**Reply by meteos77** [reply]
1st have some time to think :-)

**Reply by meteos77** [reply]
I noticed several times that the BiblioteQ windows were superimposed and I had to juggle with them by moving them
I don't know if the tabs (like in firefox) are not more adapted to your magic window (if it is adapted ?)

**Reply by meteos77** [reply]
I will reread everything you have written but I think you have made a good synthesis of the different requests

**Reply by textbrowser** [reply]
Tabs are nice for management. Windows are nice if you want to see everything.

**Comment by textbrowser** [comment]
Contains models.

https://help.ubuntu.com/community/BarcodeReaders

**Comment by textbrowser** [comment]
https://doc-snapshots.qt.io/qt6-dev/qserialport.html

**Reply by meteos77** [reply]
in serial port ?

**Reply by textbrowser** [reply]
> in serial port ?

USB = Universal Serial Bus. :)

**Reply by meteos77** [reply]
ok for usb
https://en.wikipedia.org/wiki/Serial_port   => ko : maybe for the old ones :-)

**Reply by textbrowser** [reply]
And some older scanners have RS-232 interfaces. I'll be attempting to abstract the hardware interface away by allowing the person to select the device file (think /dev). There will be a configuration panel for describing the device as well as a test mechanism.

**Comment by textbrowser** [comment]
I'll purchase something inexpensive from eBay to begin experimenting. USB communications.

**Comment by textbrowser** [comment]
There will be a new panel where one will educate BiblioteQ about their scanner. Included will be options pertaining to the device: location, rate, etc. A test interface is also required. If BQ is able to acquire test data from the device, it'll display the results in a field.

**Reply by meteos77** [reply]
I see that you have a lot of ideas :-)
I hope that scanners are easy to integrate because my experience on (joysticks and voltage measuring devices) have shown problematic init sequences for developers

**Reply by textbrowser** [reply]
On Linux, it will be easier. Some scanners can be configured as USB-HID, or keyboards. The difficulty will be to identify the device and capture its data. Then, to have BQ allow the captured data to be modified for the device without software changes.

For Mac and Windows, other problems.

On Linux, there are also permission problems. So, there will be people expecting full plug-and-play and they'll have to satisfy their own curiosities.

**Reply by meteos77** [reply]
little reminder
for the scanner access permission problems, with the other project I had to add udev rules
create file "/etc/rules.d/50-BibliotQ.rules
KERNEL=="hidraw*", ATTRS{idVendor}=="xxxx", ATTRS{idProduct}=="yyyy", MODE="0666"
SUBSYSTEM=="tty", ATTRS{idVendor}=="zzzz", ATTRS{idProduct}=="wwww", MODE="0666"

**Comment by textbrowser** [comment]
And not all scanners are capable of identifying ISBNs. Some are point-of-sale devices. Librarians will need to educate themselves on their devices.

**Comment by textbrowser** [comment]
http://wiki.opticonusa.com/techsupport/en/Linux

**Comment by textbrowser** [comment]
When proprietary software companies provide software for libraries, they lock in their approved hardware. It's a relationship between software and hardware. Software like BQ has to be impossibly flexible and people do not realize how difficult that is.

**Comment by textbrowser** [comment]
Especially when you distribute ready-to-use bundles for Mac and Windows using computers that are ancient. Qt and Homebrew are not totally made for easy distributions. For example, Mac releases require constant updates of Homebrew dependencies. People like the easy DMG and ZIP files but there's a lot of extra work that gets them these products which Qt does not provide.

**Reply by textbrowser** [reply]
My regret is providing releases instead of software. :P

**Reply by meteos77** [reply]
my history with macs stopped around 1994-95
https://fr.wikipedia.org/wiki/Macintosh_LC :-)

**Reply by meteos77** [reply]
the packaged versions are part of the software from the user's point of view :-)
otherwise look for package owners among the users (like translations)

**Comment by textbrowser** [comment]
Purchased a new USB barcode scanner from eBay. We'll see. :)

**Comment by textbrowser** [comment]
If you're going to purchase a scanner, make sure it can read ISBNs.

**Comment by textbrowser** [comment]
[opr_3001.pdf](https://github.com/textbrowser/biblioteq/files/8069696/opr_3001.pdf)

**Reply by meteos77** [reply]
there is a technical doc :-)

**Comment by textbrowser** [comment]
Yes, I researched that brand. They are not defunct and have a page about Linux.

**Comment by textbrowser** [comment]
Testing and writing the interface for it is the only guarantee that it will actually work. If it works, others will work too.

**Reply by meteos77** [reply]
I will go to the library to see the exact model we have in our possession, I hope they will be compatible

**Reply by meteos77** [reply]
with doc

**Comment by textbrowser** [comment]
A test panel where one can view the data from the scanner will be very helpful.

**Reply by meteos77** [reply]
my own experience , on linux : plug in , open a text editor , scanner => fill in the barcode : 1 per line + line feed
no configuration required, the kernel seems to see it as a keyboard (linuxmint 20.1 or 20.2)

on windows 10 : with the library software, if we put the caps lock then characters appear instead of the barcode number.

**Comment by textbrowser** [comment]
After you click on the scanner, it reads the barcode, and transfers the data directly to the input program which has keyboard focus?

**Comment by meteos77** [comment]
on my linux laptop yes, for windows the driver was probably installed before I arrived

**Reply by meteos77** [reply]
I would look on http://www.linux-usb.org/usb.ids to see if ours is in but I don't have the info on site yet

**Reply by meteos77** [reply]
what i just understood is that your scanner can be configured by reading barcodes. it's always magic with computers :-)
http://ftp.opticonusa.com/UniversalMenuBook/UniversalMenuBook.pdf

**Reply by textbrowser** [reply]
065a  Optoelectronics Co., Ltd
	0001  Opticon OPR-2001 / NLV-1001 (keyboard mode)
	0009  NLV-1001 (serial mode) / OPN-2001 [Opticon]

**Reply by meteos77** [reply]
it's crazy the number of parameters of your model
OPR-2001 = OPR-3001 ?

**Reply by meteos77** [reply]
if not recognized I do not doubt that in some time there will be 065a:0002 for the 3001 :-)

**Reply by textbrowser** [reply]
I'm hoping the 3001 will work and that the 3001 arrived after the 2001 and that the 3001 is an improvement of the 2001 while preserving compatibility. We shall see. Also, that list is probably not inclusive of every model.

**Comment by textbrowser** [comment]
The scanner has not arrived. Was supposed to be delivered today. Maybe tomorrow. Then we can think about batching some of the work.

**Reply by meteos77** [reply]
the magic and speed of the internet and the slowness of travel
I like the world of snails :-
![escargot](https://user-images.githubusercontent.com/56431039/155475403-63d1be36-f450-4ed6-b9e9-8218d0ca5db5.jpg)
)

**Comment by textbrowser** [comment]
The scanner OPR-3001 works well in Linux and Mac. It will become my official scanner for BQ. I did not have to change the systems.

**Reply by textbrowser** [reply]
Works in Windows 11 too.

**Reply by meteos77** [reply]
happy for you that it is recognized without difficulty by the main os

**Reply by meteos77** [reply]
android ?

**Reply by textbrowser** [reply]
Does not have an input port.

**Reply by textbrowser** [reply]
I tried a bunch of items and the scanner scanned all the barcodes. Made in Japan, so of course.

**Reply by textbrowser** [reply]
$29.95 via eBay. Plus shipping. New scanners are very expensive.

**Reply by textbrowser** [reply]
This model is 12 years old, older maybe. I prefer older models as they are inexpensive and things do not end up in landfills.

**Reply by meteos77** [reply]
Better a good old machine than a new bad one :-) and long live ecology

**Comment by textbrowser** [comment]
There's a delicate balance between progress and conservation. New machines tend to require less electricity save for the energy required to create them. Older machines already exist so the cost is in their use.

**Comment by meteos77** [comment]
our model of laser barcode scanner
https://www.eskynow.com/product/Esky-USB-Automatic-Handheld-Barcode-ScannerReader-With-Free-Adjustable-Stand
